### PR TITLE
Enhancement/kotlin/retrofit2

### PR DESCRIPTION
**Pull request by Ch4rl3x**
[#5076](https://github.com/OpenAPITools/openapi-generator/issues/5076)

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### PR Description
This PR add the option to use koltin suspend functions combined with retrofit2. Also adds more complexity (authorization) and ability to customization to the ApiClient.

@jimschubert @dr4ke616 @karismann @Zomzog @andrewemery  @4brunu

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/kotlin-client/jvm-common/infrastructure/Serializer.kt.mustache [hunk 1]
```diff
@@ -35,16 +35,18 @@ import java.util.Date
 {{#nonPublicApi}}internal {{/nonPublicApi}}object Serializer {
 {{#moshi}}
     @JvmStatic
```
Comment (4brunu) [inline modules/openapi-generator/src/main/resources/kotlin-client/jvm-common/infrastructure/Serializer.kt.mustache]: @Ch4rl3x can you please explain why the moshiBuilder is lazy?

**Comment by Ch4rl3x** [comment]
@4brunu you was right. I checked it again, was my fault. moshiBuilder/gsonBuilder as lazy is not neccessary. I changed it also back to val instead of var, like before

**Comment by 4brunu** [comment]
The PR ir really big, and a bit difficult to review. 
Would it be possible to split it into smaller parts?

**Comment by Ch4rl3x** [comment]
Thats a good question. The kotlin-retrofit2 library was too simple. We tooked most parts from java-retrofit2. Because of that, its such a big PR

**Comment by 4brunu** [comment]
I saw that you changed the path of all the sample projects, and I think that's whats causing some many files being changed.
Could you please submit that change in a different PR please?

**Comment by Ch4rl3x** [comment]
Ok, so do i have to close these PR and create a clean, new one?

**Comment by 4brunu** [comment]
No, you can keep this one open.
Just create a new one with the only the current sample projects being moved around.
When the new one get's merged, the diff in this PR vs the master branch will be much smaller.

**Comment by Ch4rl3x** [comment]
@wing328 I will make a new/simpler PR without renaming etc

**Comment by wing328** [comment]
@Ch4rl3x FYI. https://github.com/OpenAPITools/openapi-generator/pull/5750 has been merged into master.

Shall we close this PR?

**Comment by Ch4rl3x** [comment]
I will make a new PR within the next few weeks hopefully

**Pull request closed without merging**